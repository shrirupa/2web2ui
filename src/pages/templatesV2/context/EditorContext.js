import React, { createContext, useEffect } from 'react';
import useRouter from 'src/hooks/useRouter';
import useEditorContent from '../hooks/useEditorContent';
import useEditorNavigation from '../hooks/useEditorNavigation';
import useEditorPreview from '../hooks/useEditorPreview';
import useEditorTabs from '../hooks/useEditorTabs';

const EditorContext = createContext();

const chainHooks = (...hooks) => (
  hooks.reduce((acc, hook) => ({ ...acc, ...hook(acc) }), {})
);

export const EditorContextProvider = ({ children, value: { getDraft, getPublished, listDomains, listSubaccounts, ...value }}) => {
  const { requestParams } = useRouter();
  const pageValue = chainHooks(
    () => value,
    useEditorContent,
    useEditorNavigation,
    useEditorPreview,
    useEditorTabs
  );

  useEffect(() => {
    getDraft(requestParams.id, requestParams.subaccount);
    getPublished(requestParams.id, requestParams.subaccount);
    listDomains();
    listSubaccounts();
  }, [listSubaccounts, listDomains, getDraft, getPublished, requestParams.id, requestParams.version, requestParams.subaccount]);

  return (
    <EditorContext.Provider value={pageValue}>
      {children}
    </EditorContext.Provider>
  );
};

export default EditorContext;
