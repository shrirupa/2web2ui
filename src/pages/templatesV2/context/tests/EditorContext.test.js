import React from 'react';
import { mount, shallow } from 'enzyme';
import useRouter from 'src/hooks/useRouter';
import { EditorContextProvider } from '../EditorContext';

jest.mock('src/hooks/useRouter');

describe('EditorContext', () => {
  describe('EditorContextProvider', () => {
    const subject = ({ render = shallow, value = {}} = {}) => {
      useRouter.mockReturnValue({ requestParams: { id: 'test-template', subaccount: '123' }});

      return render(
        <EditorContextProvider
          value={{ getDraft: () => {}, getPublished: () => {}, listDomains: () => {}, listSubaccounts: () => {}, ...value }}
        >
          <div>Hello</div>
        </EditorContextProvider>
      );
    };

    it('renders children wrapped by a context provider', () => {
      expect(subject()).toMatchSnapshot();
    });

    it('sets provider value from props', () => {
      const value = { test: 'example' };
      const wrapper = subject({ value });

      expect(wrapper).toHaveProp('value', expect.objectContaining(value));
    });

    it('calls getDraft and getPublished on mount', () => {
      const getDraft = jest.fn();
      const getPublished = jest.fn();

      subject({
        render: mount, // for useEffect
        value: { getDraft, getPublished, getTestDataFromLocalStorage: jest.fn() }
      });

      expect(getDraft).toHaveBeenCalledWith('test-template', '123');
      expect(getPublished).toHaveBeenCalledWith('test-template', '123');
    });
  });
});
