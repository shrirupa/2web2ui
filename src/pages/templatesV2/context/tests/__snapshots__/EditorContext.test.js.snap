// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditorContext EditorContextProvider renders children wrapped by a context provider 1`] = `
<ContextProvider
  value={
    Object {
      "annotations": Object {
        "amp_html": Array [],
        "html": Array [],
        "text": Array [],
      },
      "content": Object {},
      "currentNavigationIndex": 0,
      "currentNavigationKey": "content",
      "currentTabIndex": 0,
      "currentTabKey": "html",
      "formattedTestData": Object {},
      "previewDevice": "desktop",
      "setContent": [Function],
      "setNavigation": [Function],
      "setPreviewDevice": [Function],
      "setTab": [Function],
      "setTestData": [Function],
      "syncTestDataInStorage": [Function],
      "testData": undefined,
    }
  }
>
  <div>
    Hello
  </div>
</ContextProvider>
`;
